File: apps/coach-web/src/lib/units.ts
export type Unit = 'mi' | 'km'
const KM_PER_MI = 1.609344

/** Stored miles → display value in the chosen unit. */
export function fromMiles(miles: number, unit: Unit): number {
  return unit === 'km' ? miles * KM_PER_MI : miles
}
/** Display value in the unit → miles (for storage). */
export function toMiles(value: number, unit: Unit): number {
  return unit === 'km' ? value / KM_PER_MI : value
}
/** Format a stored mileage for display (number only, no unit label). */
export function fmtDist(miles: number | null | undefined, unit: Unit): string {
  if (miles == null) return ''
  const v = fromMiles(miles, unit)
  return v % 1 ? v.toFixed(1) : String(v)
}
/** Convert a stored pace string "M:SS/mi" to the chosen unit, e.g. "5:17/km". */
export function fmtPace(pace: string | null | undefined, unit: Unit): string {
  if (!pace) return ''
  const m = pace.match(/(\d+):(\d{1,2})/)
  if (!m) return pace
  let sec = Number(m[1]) * 60 + Number(m[2]) // seconds per mile
  if (unit === 'km') sec = Math.round(sec / KM_PER_MI)
  return `${Math.floor(sec / 60)}:${String(sec % 60).padStart(2, '0')}/${unit}`
}

File: apps/athlete-ios/RecBuddy/Lib/Units.swift
import Foundation

enum Unit: String { case mi, km }

/// mi/km display conversion — port of coach-web's lib/units.ts.
/// Storage is ALWAYS miles + "M:SS/mi"; unit only changes display.
enum Units {
    static let kmPerMi = 1.609344

    static func fromMiles(_ miles: Double, _ unit: Unit) -> Double {
        unit == .km ? miles * kmPerMi : miles
    }
    static func toMiles(_ value: Double, _ unit: Unit) -> Double {
        unit == .km ? value / kmPerMi : value
    }
    /// Stored mileage -> display number ("6", "4.5", "16.1"); "" for nil.
    /// Mirrors TS `v % 1 ? v.toFixed(1) : String(v)` — the whole-number check is
    /// on the RAW value, so 9.977 km displays "10.0" (rounded), not "10".
    static func fmtDist(_ miles: Double?, _ unit: Unit) -> String {
        guard let miles else { return "" }
        let v = fromMiles(miles, unit)
        return v.truncatingRemainder(dividingBy: 1) == 0
            ? String(Int(v))
            : String(format: "%.1f", v)
    }
    /// Stored "M:SS/mi" -> display pace in the unit, e.g. "5:17/km". "" for nil.
    static func fmtPace(_ pace: String?, _ unit: Unit) -> String {
        guard let pace else { return "" }
        guard var sec = Pace.toSeconds(pace) else { return pace }
        if unit == .km { sec = Int((Double(sec) / kmPerMi).rounded()) }
        return "\(sec / 60):" + String(format: "%02d", sec % 60) + "/\(unit.rawValue)"
    }
}
